src/Koffee/Utility.fs

```diff
@@ -76,6 +76,16 @@ type AsyncSeqResultBuilder() =
 
 let asyncSeqResult = AsyncSeqResultBuilder()
 
+module AsyncSeq =
+    let repeatResult times (seqBuilder: 'a -> AsyncSeq<Result<'a,'b>>) initial = asyncSeq {
+        let mutable state = initial
+        for _ in [1..times] do
+            yield! (seqBuilder(state)
+                    |> AsyncSeq.map (function
+                        | Ok newState -> state <- newState; Ok newState
+                        | res -> res))
+    }
```

Comment:
@mattstermiller Any idea on how to accomplish the same thing I'm trying to do here, by without `mutable`?

Reply:
Not really. I'll let you in on a little secret: VinylUI has a `mutable currentModel` inside.

Ideally, we wouldn't need this function at all and implement repetition in each handler that supports it. Back and forward would benefit since there would be no need to load directory listings for intermediate paths. For example, 5H would move 5 items off of the back stack and only load the last one. However, if this works, I am willing to accept for now and create a new issue for that optimization.

I think this implementation has a problem though. If an error is encountered, it should stop instead of continuing. The asyncSeqResult builder has this behavior - if an Error value is encountered, it stops and does not produce any more values.

Reply:
Hehe yea, trying to get into the "mutability is evil" mindset, but maybe sometimes it's ok. The mutable state here doens't leek and is only used inside this function, so I guess that's what clears it?

This function is only used in the undo'ing and redo'ing commands. Stepping back and forth by count actually does a full multiple steps jump, instead of incrementing. There's an issue with that, with that if the destination path in the "Back 5" command does not exist, it will fail the jump altogether, instead of jumping back 4 times then, or trying to jump back to the next one if that one exists.

The error handling is a very good catch! Totally forgot about it!

Reply:
I was able to implement what you suggested. Bonus was that it was quite easy to add a way to show the progress when the undo/redo functions dealt with the repeating themselves. Here's an example, where I added a `let! _ = Async.Wait(Task.Delay 1000)` just to emphasize it:

![koffee-repeat-undo-progress](https://user-images.githubusercontent.com/2477952/101259867-de70fc00-372b-11eb-8506-55cda3a33146.gif)

I just moved the entire undo and redo implementations into a recursive function, then called that in the true undo/redo functions. Maybe the repeating can be abstracted away somehow. What do you think?

Reply:
I left a review before seeing this comment. I like your recursive implementation, but I don't think the status updates would actually work out the way you're thinking. To simulate the undo operations taking some time, you would need to add the delay where the controller is calling into the filesystem. Each undo operation sets the status to a Busy state to show that it's doing something as well as "lock" the UI so that the user can't try to do something else simultaneously. I said in a review comment, these intermediate "2 of 4 done" messages would be immediately replaced by a Busy status and would never actually be seen.